src/wiktextract/extractor/en/analyze_template.py

```diff
@@ -0,0 +1,138 @@
+
+    # Determine if the template contains table element tokens
+    # outside paired table start/end.  We only try to look for
+    # these outside templates, as it is common to write each
+    # template argument on its own line starting with a "|".
+    outside = unpaired_text
+    while True:
+        # print("=== OUTSIDE ITER")
+        prev = outside
+
+        # handle {{{ }}} parameters without templates inside them
+        while True:
+            newt = re.sub(
+                # re.X, ignore white space and comments
+                r"""(?sx)\{\{\{                # {{{
+                               (    [^{}]      # no {} except...
+                               |    \}[^}]     # no }} unless...
+                               |    \}\}[^}]   # they're definitely not }}}
+                               )*?
+                         \}\}\}                # }}}
+                """,
+                "",
+                prev,
+            )
+            if newt == prev:
+                break
+            prev = newt
+        # print("After arg elim: {!r}".format(newt))
+
+        # Handle templates
+        newt = re.sub(
+            r"""(?sx)\{\{
+                                    (    [^{}]
+                                    |    \}[^}]
+                                    )*?
+                                \}\}""",
+            "",
+            newt,
+        )
+        # print("After templ elim: {!r}".format(newt))
+        if newt == outside:
+            break
+        outside = newt
+    # Check if the template contains certain table elements
+    # start of line plus |+, |- or |!
+    m = re.search(r"(?s)(^|\n)(\|\+|\|-|\!)", outside)
+    m2 = re.match(r"(?si)\s*(<includeonly>|<!--.*?-->)(\|\||!!)", outside)
+    contains_table_element = m is not None or m2 is not None
+    # if contains_table_element:
+    #     print("contains_table_element {!r} at {}"
+    #           .format(m.group(0), m.start()))
+    #     print("... {!r} ...".format(outside[m.start() - 10:m.end() + 10]))
+    #     print(repr(outside))
+
+    # Check for unpaired HTML tags
+    tag_cnts: defaultdict[str, int] = defaultdict(int)
+    for m in re.finditer(
+        r"(?si)<(/)?({})\b\s*[^>]*(/)?>" r"".format(
+            "|".join(wtp.paired_html_tags)
+        ),
+        outside,
+    ):
+        start_slash = m.group(1)
+        tagname = m.group(2)
+        end_slash = m.group(3)
+        if start_slash:
+            tag_cnts[tagname] -= 1
+        elif not end_slash:
+            tag_cnts[tagname] += 1
+    contains_unbalanced_html = any(v != 0 for v in tag_cnts.values())
+    # if contains_unbalanced_html:
+    #     print(name, "UNBALANCED HTML")
+    #     for k, v in tag_cnts.items():
+    #         if v != 0:
+    #             print("  {} {}".format(v, k))
+
+    return (
+        contains_list
+        or contains_unpaired_table
+        or contains_table_element
+        or contains_unbalanced_html
+    )
```

Comment:
Instead of `or`ing here, I think this could all be technically done with early returns.

Reply:
I didn't think about that... I mostly want minimal changes to these complex code for easier compare.